source/source.go

```diff
@@ -107,30 +153,51 @@ func (s *Source) Open(ctx context.Context, sdkPos opencdc.Position) (err error)
 	return nil
 }
 
-func (s *Source) Read(ctx context.Context) (rec opencdc.Record, err error) {
+func (s *Source) receiveMessage(ctx context.Context) (msg types.Message, err error) {
+	if msg, ok := s.savedMessages.nextSavedMessage(); ok {
+		return msg, nil
+	}
+
 	receiveMessage := &sqs.ReceiveMessageInput{
 		MessageAttributeNames: []string{
 			string(types.QueueAttributeNameAll),
 		},
 		QueueUrl:            &s.queueURL,
-		MaxNumberOfMessages: 1,
+		MaxNumberOfMessages: s.config.MaxNumberOfMessages,
 		VisibilityTimeout:   s.config.VisibilityTimeout,
 		WaitTimeSeconds:     s.config.WaitTimeSeconds,
 	}
 
-	// grab a message from queue
 	sqsMessages, err := s.svc.ReceiveMessage(ctx, receiveMessage)
 	if err != nil {
-		return rec, fmt.Errorf("error retrieving amazon sqs messages: %w", err)
+		return msg, fmt.Errorf("error retrieving amazon sqs messages: %w", err)
 	}
-
-	// if there are no messages in queue, backoff
 	if len(sqsMessages.Messages) == 0 {
 		sdk.Logger(ctx).Warn().Msg("got 0 messages from queue")
-		return rec, sdk.ErrBackoffRetry
+		return msg, sdk.ErrBackoffRetry
+	}
+
+	if s.receiveMessageCalled != nil {
+		s.receiveMessageCalled()
+	}
+
+	if len(sqsMessages.Messages) == 1 {
+		return sqsMessages.Messages[0], nil
 	}
 
-	message := sqsMessages.Messages[0]
+	// While we wait for ReceiveMessage to return, another concurrent `source.Read`
```

Comment:
`Source.Read` is never called concurrently, Conduit always calls it sequentially, guaranteed. This means we can simplify this, no need for locks etc.